mindsdb/interfaces/skills/sql_agent.py

```diff
@@ -117,18 +133,61 @@ def get_table_info(self, table_names: Optional[List[str]] = None) -> str:
         If `sample_rows_in_table_info`, the specified number of sample rows will be
         appended to each table description. This can increase performance as demonstrated in the paper.
         """
+        try:
+            cache_key = self._generate_cache_key(table_names)
```

Comment:
How often the llm use more than 1 table in get_table_info?
If there is 1 table in 99% cases - we could cache per table  (and simplify generation key and update of cache)

Reply:
so the issue is the agent will identify which table(s) should be used from the list of tables that are in a DB and only will get the schemas for these tables.

Therefore, if we only use DB_name as cache key, then sometimes it will not contain all the tables.


I saw this happen quite a few times, it was significant for sure.

Reply:
example:


```
New input: Answer the user"s question in a helpful way: what is the average miles per gallon for the cars in the database?
Thought: Do I need to use a tool? Yes
Action: sql_db_list_tables
Action Input: ""
Observation: mindsdb_sql_agent_datasource.cta, mindsdb_sql_agent_datasource.car_info, mindsdb_sql_agent_datasource.car_sales, mindsdb_sql_agent_datasource.financial_headlines, mindsdb_sql_agent_datasource.home_rentals, mindsdb_sql_agent_datasource.iris, mindsdb_sql_agent_datasource.amazon_reviews, mindsdb_sql_agent_datasource.review_sentiment, mindsdb_sql_agent_datasource.chat_llm_mindsdb_docs, mindsdb_sql_agent_datasource.house_sales, mindsdb_sql_agent_datasource.used_car_price, mindsdb_sql_agent_datasource.fraud_detection, mindsdb_sql_agent_datasource.customer_churn, mindsdb_sql_agent_datasource.bank_customer_transactions, mindsdb_sql_agent_datasource.user_comments, mindsdb_sql_agent_datasource.customer_support_chat, mindsdb_sql_agent_datasource.telecom_customer_churn
Thought:The user wants to know the average miles per gallon for the cars in the database. From the list of tables, the most likely table to contain this information is "car_info" or "car_sales". I need to check the schema of these tables to confirm if they have a column for miles per gallon. 
Do I need to use a tool? Yes
Action: sql_db_schema
Action Input: car_info, car_sales
Observation: Table named `car_info`
/* Sample with first 3 rows from table mindsdb_sql_agent_datasource.car_info:
model	year	transmission	fueltype	mpg	enginesize
 A1	2010	Automatic	Petrol	53.3	1.4
 A1	2011	Manual	Diesel	70.6	1.6
 A1	2011	Manual	Petrol	53.3	1.4
*/
Column data types: `model` : `text`,	`year` : `integer`,	`transmission` : `text`,	`fueltype` : `text`,	`mpg` : `double precision`,	`enginesize` : `double precision`
Table named `car_sales`
/* Sample with first 3 rows from table mindsdb_sql_agent_datasource.car_sales:
model	year	price	transmission	mileage	fueltype	tax
 A1	2017	12500	Manual	15735	Petrol	150
 A6	2016	16500	Automatic	36203	Diesel	20
 A1	2016	11000	Manual	29946	Petrol	30
*/
Column data types: `model` : `text`,	`year` : `integer`,	`price` : `integer`,	`transmission` : `text`,	`mileage` : `integer`,	`fueltype` : `text`,	`tax` : `integer`
Thought:The 'car_info' table contains a column 'mpg' which likely stands for miles per gallon. I can use this table to calculate the average miles per gallon for the cars in the database. I will write a SQL query to calculate this average and then use the sql_db_query tool to execute this query.
Do I need to use a tool? Yes
Action: mindsdb_sql_parser_tool
Action Input: SELECT AVG(mpg) as average_mpg FROM car_info
Observation: valid query: SELECT avg(mpg) AS average_mpg FROM car_info
```